File: Utility/XML/HTML/modURLHelper.vb
Option Explicit On 

Structure URL
    Dim Scheme As String
    Dim Host As String
    Dim Port As Long
    Dim URI As String
    Dim Query As QueryString()
End Structure

Structure QueryString
    Dim Name As String
    Dim Value As String
End Structure

Module URLHelper

    ' returns as type URL from a string
    Function ExtractUrl(ByVal strUrl As String) As URL
        Dim intPos1 As Integer
        Dim intPos2 As Integer

        Dim retURL As New URL

        '1 look for a scheme it ends with ://
        intPos1 = InStr(strUrl, "://")

        If intPos1 > 0 Then
            retURL.Scheme = Mid(strUrl, 1, intPos1 - 1)
            strUrl = Mid(strUrl, intPos1 + 3)
        End If

        '2 look for a port
        intPos1 = InStr(strUrl, ":")
        intPos2 = InStr(strUrl, "/")

        If intPos1 > 0 And intPos1 < intPos2 Then
            ' a port is specified
            retURL.Host = Mid(strUrl, 1, intPos1 - 1)

            If (IsNumeric(Mid(strUrl, intPos1 + 1, intPos2 - intPos1 - 1))) Then
                retURL.Port = CInt(Mid(strUrl, intPos1 + 1, intPos2 - intPos1 - 1))
            End If
        ElseIf intPos2 > 0 Then
            retURL.Host = Mid(strUrl, 1, intPos2 - 1)
        Else
            retURL.Host = strUrl
            retURL.URI = "/"

            ExtractUrl = retURL
            Exit Function
        End If

        strUrl = Mid(strUrl, intPos2)

        ' find a question mark ?
        intPos1 = InStr(strUrl, "?")

        If intPos1 > 0 Then
            retURL.URI = Mid(strUrl, 1, intPos1 - 1)
            retURL.Query = SplitParms(Mid(strUrl, intPos1 + 1))
        Else
            retURL.URI = strUrl
        End If

        ExtractUrl = retURL
    End Function

    Function SplitParms(ByVal Query As String) As QueryString()

        Dim idx As Integer
        Dim parms As String()
        Dim value As String()
        Dim ret As QueryString()

        parms = Split(Query, "&")

        ReDim ret(UBound(parms))

        For idx = 0 To UBound(parms)

            value = Split(parms(idx), "=")

            ret(idx).Name = value(0)
            ret(idx).Value = value(1)

        Next

        Return ret

    End Function

    ' url encodes a string
    Function URLEncode(ByVal str As String) As String
        Dim intLen As Integer
        Dim X As Integer
        Dim curChar As Long
        Dim newStr As String
        intLen = Len(Str)
        newStr = ""
        For X = 1 To intLen
            curChar = Asc(Mid$(Str, X, 1))

            If (curChar < 48 Or curChar > 57) And _
                (curChar < 65 Or curChar > 90) And _
                (curChar < 97 Or curChar > 122) Then
                newStr = newStr & "%" & Hex(curChar)
            Else
                newStr = newStr & Chr(curChar)
            End If
        Next X

        URLEncode = newStr
    End Function

    ' decodes a url encoded string
    Function UrlDecode(ByVal str As String) As String
        Dim intLen As Integer
        Dim X As Integer
        Dim curChar As String
        Dim strCode As String

        Dim newStr As String

        intLen = Len(str)
        newStr = ""

        For X = 1 To intLen
            curChar = Mid$(str, X, 1)

            If curChar = "%" Then
                strCode = "&h" & Mid$(str, X + 1, 2)

                If IsNumeric(strCode) Then
                    curChar = Chr(Int(strCode))
                Else
                    curChar = ""
                End If
                X = X + 2
            End If

            newStr = newStr & curChar
        Next X

        UrlDecode = newStr
    End Function

End Module

File: FFDownload/Utility/Browser/HTML/modURLHelper.cs
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace HtmlGrabber
{

    struct URL
    {
        public string Scheme;
        public string Host;
        public long Port;
        public string URI;
        public QueryString[] Query;
    }

    struct QueryString
    {
        public string Name;
        public string Value;
    }

    static class URLHelper
    {

        // returns as type URL from a string
        public static URL ExtractUrl(string strUrl)
        {
            URL ExtractUrlRet = default;
            int intPos1;
            int intPos2;

            var retURL = new URL();

            // 1 look for a scheme it ends with ://
            intPos1 = Strings.InStr(strUrl, "://");

            if (intPos1 > 0)
            {
                retURL.Scheme = Strings.Mid(strUrl, 1, intPos1 - 1);
                strUrl = Strings.Mid(strUrl, intPos1 + 3);
            }

            // 2 look for a port
            intPos1 = Strings.InStr(strUrl, ":");
            intPos2 = Strings.InStr(strUrl, "/");

            if (intPos1 > 0 & intPos1 < intPos2)
            {
                // a port is specified
                retURL.Host = Strings.Mid(strUrl, 1, intPos1 - 1);

                if (Information.IsNumeric(Strings.Mid(strUrl, intPos1 + 1, intPos2 - intPos1 - 1)))
                {
                    retURL.Port = Conversions.ToInteger(Strings.Mid(strUrl, intPos1 + 1, intPos2 - intPos1 - 1));
                }
            }
            else if (intPos2 > 0)
            {
                retURL.Host = Strings.Mid(strUrl, 1, intPos2 - 1);
            }
            else
            {
                retURL.Host = strUrl;
                retURL.URI = "/";

                ExtractUrlRet = retURL;
                return ExtractUrlRet;
            }

            strUrl = Strings.Mid(strUrl, intPos2);

            // find a question mark ?
            intPos1 = Strings.InStr(strUrl, "?");

            if (intPos1 > 0)
            {
                retURL.URI = Strings.Mid(strUrl, 1, intPos1 - 1);
                retURL.Query = SplitParms(Strings.Mid(strUrl, intPos1 + 1));
            }
            else
            {
                retURL.URI = strUrl;
            }

            ExtractUrlRet = retURL;
            return ExtractUrlRet;
        }

        public static QueryString[] SplitParms(string Query)
        {

            int idx;
            string[] parms;
            string[] value;
            QueryString[] ret;

            parms = Strings.Split(Query, "&");

            ret = new QueryString[Information.UBound(parms) + 1];

            var loopTo = Information.UBound(parms);
            for (idx = 0; idx <= loopTo; idx++)
            {

                value = Strings.Split(parms[idx], "=");

                ret[idx].Name = value[0];
                ret[idx].Value = value[1];

            }

            return ret;

        }

        // url encodes a string
        public static string URLEncode(string str)
        {
            string URLEncodeRet = default;
            int intLen;
            int X;
            long curChar;
            string newStr;
            intLen = Strings.Len(str);
            newStr = "";
            var loopTo = intLen;
            for (X = 1; X <= loopTo; X++)
            {
                curChar = Strings.Asc(Strings.Mid(str, X, 1));

                if ((curChar < 48L | curChar > 57L) & (curChar < 65L | curChar > 90L) & (curChar < 97L | curChar > 122L))

                {
                    newStr = newStr + "%" + Conversion.Hex(curChar);
                }
                else
                {
                    newStr = newStr + Strings.Chr((int)curChar);
                }
            }

            URLEncodeRet = newStr;
            return URLEncodeRet;
        }

        // decodes a url encoded string
        public static string UrlDecode(string str)
        {
            string UrlDecodeRet = default;
            int intLen;
            int X;
            string curChar;
            string strCode;

            string newStr;

            intLen = Strings.Len(str);
            newStr = "";

            var loopTo = intLen;
            for (X = 1; X <= loopTo; X++)
            {
                curChar = Strings.Mid(str, X, 1);

                if (curChar == "%")
                {
                    strCode = "&h" + Strings.Mid(str, X + 1, 2);

                    if (Information.IsNumeric(strCode))
                    {
                        curChar = Conversions.ToString(Strings.Chr(Conversions.ToInteger(Conversion.Int(strCode))));
                    }
                    else
                    {
                        curChar = "";
                    }
                    X = X + 2;
                }

                newStr = newStr + curChar;
            }

            UrlDecodeRet = newStr;
            return UrlDecodeRet;
        }

    }
}